docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+__String literal__ will support the same Prometheus match type. (`=`,`=!`,`!~`,`=~`).
+
+Example: `{job="prod/query-frontend"} | logfmt | filter level=~"info|debug" and query_type!="filter"`
+
+__Numeric literal__ will support `=`,`!=` but also `<`,`<=`,`>`,`>=`.
+
+Example: `{job="prod/query-frontend"} | json | filter latency > 10 or status >= 400`
+
+When field value and literal types are different, conversion will be attempted prior testing and will be translated into zero type values in case of failure.
+
+In the future we will add more literals such as durations `2s` and dates (`03-03-2020T20:30`).
+
+### Sort
+
+The sort operation noted `| sort` allows to orders fields entries by a field value instead of the field entry time.
+
+> It should be noted that sorting is always executed after parsing fields entries in the time range. Not on the whole time range.
+
+For example: `{job="prod/query-frontend"} | json | sort latency desc`
+
+will sort  the first 1000 (default limit) fields entries by the highest latency first. Multiple fields can be used as fallback by the sort operation in case of equality.
+
+For example: `{job="prod/query-frontend"} | json | sort latency desc path asc` will sort by latency descending first then path ascending.
+
+Using sort operator with metric queries will result into an error.(e.g `count_over_time({job="prod/query-frontend"} | logfmt  | sort latency desc [5m])`)
+
+> __Warning:__ Sorting is dependent on how Grafana is currently ordering lines. Even if we re order our responses there is no guarantee that Grafana will be relying on it. And we might have to remove uncommon labels to keep only stream to ensure ordering across field streams. This means we will need a new API (`/v2`) to support ordering, although with compromise we might get something to work with the `v1` API.
+
+### Select
+
+The Select operator (`| select`) allows you to select and reduce field that should be returned.
+
+If you don't provide a select operator, all fields parsed will be returned by default. This means all fields from a log line if you're using implicit extraction. However if you used explicit extraction, only those fields will be returned.
+
+For example this query:
+
+```logql
+{job="prod/query-frontend"} | logfmt | sort duration desc | limit 25 | select query duration
+```
+
+is the same as this one
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+```
+
+because the field extraction was explicit, only those fields (`query` and `duration`) are outputted.
+
+This is why select is important to use for large queries. Select will allows users to remove data (column) in their logs that are useless, repetitive and not interesting for the current investigation.
+
+The select operation will also support the [go templating language](https://golang.org/pkg/text/template/) when supplied with string argument and not fields.
```

Comment:
A little concerned about exposing all of this functionality.  It's a large surface area for both security and performance issues.

Reply:
Fair point.

I expect this to come at a later stage and I think we should be careful.

Reply:
To me, the select operator looks the wrong operator to add templating support. I would expect the select operator to just select fields and not transform them. I'm personally dubious about supporting the go templating language at all in any operator but if so I would suggest to use a dedicated operator.

Reply:
Good feedback !

I hesitate on this. I think we want to allow re-formatting in case the data is not visually helpful or meaningful.

Based on @owen-d feedback, I'm wondering if `select` is really need and if we just don't want a formatter operator.

The uses cases was to reduce noise from logs that are redundant, basically covering pipeparse use case from @slim-bean